File: plugins/task-gate/scripts/codex_gate.py
#!/usr/bin/env python3
from __future__ import annotations

import argparse
import json
import os
import re
import subprocess
import sys
from dataclasses import dataclass, field
from pathlib import Path
from typing import Any, Callable


PLUGIN_ROOT = Path(__file__).resolve().parents[1]
if str(PLUGIN_ROOT) not in sys.path:
    sys.path.insert(0, str(PLUGIN_ROOT))

from scripts.task_gate import (  # noqa: E402
    PLAN_SCHEMA,
    PlanError,
    Task,
    TaskPlan,
    TaskPlanner,
    Thinker,
    build_default_thinker,
)


FOLLOWUP_SCHEMA: dict[str, Any] = {
    "type": "object",
    "properties": {
        "complete": {"type": "boolean"},
        "summary": {"type": "string"},
        "next_tasks": PLAN_SCHEMA["properties"]["tasks"],
    },
    "required": ["complete", "summary"],
    "additionalProperties": False,
}


@dataclass(frozen=True)
class GateResult:
    exit_code: int
    output: str = ""


@dataclass(frozen=True)
class CodexRunResult:
    exit_code: int
    output: str = ""


@dataclass(frozen=True)
class FollowupDecision:
    complete: bool
    summary: str = ""
    next_tasks: list[Task] = field(default_factory=list)


class CodexExecutor:
    def __init__(
        self,
        command: str | None = None,
        runner: Callable[..., subprocess.CompletedProcess[str]] = subprocess.run,
    ) -> None:
        self.command = command or os.environ.get("TASK_GATE_CODEX_BIN", "codex")
        self.runner = runner

    def execute(
        self,
        codex_prompt: str,
        cwd: str | None = None,
        extra_args: list[str] | None = None,
    ) -> CodexRunResult:
        args = [self.command, "exec", "--skip-git-repo-check"]
        if cwd:
            args.extend(["-C", cwd])
        if extra_args:
            args.extend(extra_args)
        args.append(codex_prompt)

        try:
            completed = self.runner(args, text=True, check=False, capture_output=True)
        except FileNotFoundError as exc:
            raise PlanError(f"codex command not found: {self.command}") from exc

        stdout = (getattr(completed, "stdout", "") or "").strip()
        stderr = (getattr(completed, "stderr", "") or "").strip()
        output = "\n".join(part for part in [stdout, stderr] if part)
        return CodexRunResult(
            exit_code=int(getattr(completed, "returncode", 0)),
            output=output,
        )


class FollowupPlanner:
    def __init__(self, thinker: Thinker | None = None, max_tasks: int = 8) -> None:
        self.thinker = thinker or build_default_thinker(FOLLOWUP_SCHEMA)
        self.max_tasks = max_tasks

    def assess(
        self,
        *,
        plan: TaskPlan,
        codex_output: str,
        exit_code: int,
        round_number: int,
    ) -> FollowupDecision:
        prompt = build_followup_prompt(
            plan=plan,
            codex_output=codex_output,
            exit_code=exit_code,
            round_number=round_number,
            max_tasks=self.max_tasks,
        )
        output = self.thinker.think(prompt)
        return parse_followup_output(output, max_tasks=self.max_tasks)


class CodexGate:
    def __init__(
        self,
        planner: TaskPlanner | None = None,
        executor: CodexExecutor | None = None,
        followup_planner: FollowupPlanner | None = None,
    ) -> None:
        self.planner = planner or TaskPlanner()
        self.executor = executor or CodexExecutor()
        self.followup_planner = followup_planner or FollowupPlanner()

    def run(
        self,
        prompt: str,
        execute: bool = False,
        cwd: str | None = None,
        codex_args: list[str] | None = None,
        max_rounds: int = 3,
    ) -> GateResult:
        try:
            plan = self.planner.plan(prompt)
        except PlanError as exc:
            return GateResult(exit_code=1, output=f"task-gate: {exc}")

        if not execute:
            return GateResult(exit_code=0, output=plan.as_numbered_text())

        if max_rounds < 1:
            return GateResult(exit_code=1, output="task-gate: max_rounds must be at least 1")

        current_plan = plan
        reports = ["Initial task plan:", plan.as_numbered_text()]
        for round_number in range(1, max_rounds + 1):
            codex_prompt = build_codex_execution_prompt(
                current_plan,
                round_number=round_number,
            )
            try:
                run_result = self.executor.execute(
                    codex_prompt,
                    cwd=cwd,
                    extra_args=codex_args,
                )
                decision = self.followup_planner.assess(
                    plan=current_plan,
                    codex_output=run_result.output,
                    exit_code=run_result.exit_code,
                    round_number=round_number,
                )
            except PlanError as exc:
                reports.append(f"task-gate: {exc}")
                return GateResult(exit_code=1, output="\n\n".join(reports))

            reports.append(_format_round_report(round_number, run_result, decision))
            if decision.complete and run_result.exit_code == 0:
                reports.append("Task Gate completion verdict: complete.")
                return GateResult(exit_code=0, output="\n\n".join(reports))

            if decision.complete and run_result.exit_code != 0:
                reports.append(
                    "Task Gate marked the work complete, but Codex exited nonzero; "
                    "continuing because completion is not proven."
                )

            if not decision.next_tasks:
                reports.append(
                    "task-gate: task is not complete and Gate returned no next tasks."
                )
                return GateResult(exit_code=1, output="\n\n".join(reports))

            current_plan = TaskPlan(
                source_prompt=f"Gate follow-up after execution round {round_number}",
                tasks=decision.next_tasks,
            )

        reports.append(
            "task-gate: max execution rounds reached before completion; "
            "refusing to report success while work remains."
        )
        return GateResult(exit_code=1, output="\n\n".join(reports))


def build_codex_execution_prompt(plan: TaskPlan, round_number: int = 1) -> str:
    return (
        "Task Gate has already converted the user's raw request into the "
        "authorized task plan below. Execute only these numbered tasks in order.\n"
        f"This is execution round {round_number}.\n\n"
        f"{_format_execution_plan(plan)}\n\n"
        "Execution rules:\n"
        "- Treat the numbered plan as the task boundary.\n"
        "- Do not reinterpret the original raw request; it is intentionally absent.\n"
        "- If a required task is missing or unsafe, stop and report the blocker.\n"
        "- Before claiming completion, run suitable verification and report evidence.\n"
        "- At the end of your response, always include a Detailed completion summary.\n\n"
        "Detailed completion summary requirements:\n"
        "- Work completed: list the concrete tasks handled and files or commands involved.\n"
        "- Verification: list commands, checks, and exact results, or say why they were not run.\n"
        "- Remaining work: list any unfinished task, missing verification, or uncertainty.\n"
        "- Blockers: list blockers or write none.\n"
        "- Completion verdict: write complete only when all authorized tasks and verification are done; otherwise write incomplete.\n"
        "This detailed summary is mandatory because Task Gate will read it to decide the next tasks."
    )


def build_followup_prompt(
    *,
    plan: TaskPlan,
    codex_output: str,
    exit_code: int,
    round_number: int,
    max_tasks: int = 8,
) -> str:
    return (
        "You are Task Gate reviewing the end of a Codex execution round. "
        "The raw user prompt is intentionally absent; use only the authorized "
        "task plan and Codex's detailed completion summary.\n"
        "Return only JSON matching this shape: "
        '{"complete":false,"summary":"detailed assessment",'
        '"next_tasks":[{"title":"next task","details":"optional detail",'
        '"acceptance_criteria":["optional check"]}]}.\n'
        f"If complete is false, provide 1 to {max_tasks} concrete next_tasks "
        "so Codex can continue. If complete is true, next_tasks may be empty. "
        "Do not mark complete when verification is missing, Codex exited nonzero, "
        "or the summary says work remains.\n\n"
        f"Execution round: {round_number}\n"
        f"Codex exit code: {exit_code}\n\n"
        "Authorized task plan:\n"
        f"{_format_execution_plan(plan)}\n\n"
        "Codex detailed completion summary and output:\n"
        f"{_clip_text(codex_output)}"
    )


def parse_followup_output(output: str, max_tasks: int = 8) -> FollowupDecision:
    parsed = _try_parse_json(output)
    if parsed is None:
        raise PlanError("follow-up gate returned invalid JSON")
    return _followup_from_json(parsed, max_tasks=max_tasks)


def _followup_from_json(parsed: Any, max_tasks: int) -> FollowupDecision:
    if isinstance(parsed, dict) and isinstance(parsed.get("structured_output"), dict):
        return _followup_from_json(parsed["structured_output"], max_tasks=max_tasks)

    if isinstance(parsed, dict) and isinstance(parsed.get("result"), str):
        nested = _try_parse_json(parsed["result"])
        if nested is not None:
            return _followup_from_json(nested, max_tasks=max_tasks)

    if not isinstance(parsed, dict):
        raise PlanError("follow-up gate JSON must be an object")

    complete = _bool_from_raw(
        parsed.get("complete")
        if "complete" in parsed
        else parsed.get("is_complete", parsed.get("done"))
    )
    summary = str(parsed.get("summary") or parsed.get("assessment") or "").strip()
    raw_tasks = parsed.get("next_tasks") or parsed.get("tasks") or []

    next_tasks: list[Task] = []
    if isinstance(raw_tasks, list):
        for raw in raw_tasks[:max_tasks]:
            task = _task_from_raw(raw, len(next_tasks) + 1)
            if task is not None:
                next_tasks.append(task)

    return FollowupDecision(complete=complete, summary=summary, next_tasks=next_tasks)


def _try_parse_json(output: str) -> Any | None:
    text = output.strip()
    fenced = re.search(r"```(?:json)?\s*(.*?)```", text, flags=re.DOTALL)
    if fenced:
        text = fenced.group(1).strip()
    else:
        object_match = re.search(r"\{.*\}", text, flags=re.DOTALL)
        if object_match:
            text = object_match.group(0)

    try:
        return json.loads(text)
    except json.JSONDecodeError:
        return None


def _bool_from_raw(raw: Any) -> bool:
    if isinstance(raw, bool):
        return raw
    if isinstance(raw, str):
        normalized = raw.strip().lower()
        if normalized in {"true", "yes", "done", "complete", "completed"}:
            return True
        if normalized in {"false", "no", "incomplete", "remaining", "blocked"}:
            return False
    raise PlanError('follow-up gate JSON must contain boolean "complete"')


def _task_from_raw(raw: Any, task_id: int) -> Task | None:
    if isinstance(raw, str):
        title = raw.strip()
        return Task(id=task_id, title=title) if title else None

    if not isinstance(raw, dict):
        return None

    title = str(
        raw.get("title")
        or raw.get("task")
        or raw.get("description")
        or raw.get("name")
        or ""
    ).strip()
    if not title:
        return None

    details = str(raw.get("details") or raw.get("detail") or "").strip()
    raw_checks = raw.get("acceptance_criteria") or raw.get("checks") or []
    return Task(
        id=task_id,
        title=title,
        details=details,
        acceptance_criteria=_string_list(raw_checks),
    )


def _string_list(raw: Any) -> list[str]:
    if isinstance(raw, list):
        return [str(item).strip() for item in raw if str(item).strip()]
    if isinstance(raw, str) and raw.strip():
        return [raw.strip()]
    return []


def _clip_text(text: str) -> str:
    try:
        limit = int(os.environ.get("TASK_GATE_CODEX_OUTPUT_CHARS", "12000"))
    except ValueError:
        limit = 12000
    if len(text) <= limit:
        return text
    omitted = len(text) - limit
    return f"[omitted {omitted} earlier chars]\n{text[-limit:]}"


def _format_round_report(
    round_number: int,
    run_result: CodexRunResult,
    decision: FollowupDecision,
) -> str:
    lines = [
        f"Execution round {round_number}",
        f"Codex exit code: {run_result.exit_code}",
        "Codex detailed completion summary:",
        _indent(run_result.output or "<no output>"),
        "Gate follow-up:",
        _indent(decision.summary or "<no summary>"),
    ]
    if decision.next_tasks:
        lines.append("Gate next tasks:")
        next_plan = TaskPlan(source_prompt="", tasks=decision.next_tasks)
        lines.append(_indent(next_plan.as_numbered_text()))
    return "\n".join(lines)


def _indent(text: str) -> str:
    return "\n".join(f"  {line}" if line else "" for line in text.splitlines())


def _format_execution_plan(plan: TaskPlan) -> str:
    lines: list[str] = []
    for task in plan.tasks:
        lines.append(f"{task.id}. {task.title}")
        if task.details:
            lines.append(f"   Details: {task.details}")
        if task.acceptance_criteria:
            lines.append("   Acceptance criteria:")
            for criterion in task.acceptance_criteria:
                lines.append(f"   - {criterion}")
    return "\n".join(lines)


def main(argv: list[str] | None = None) -> int:
    parser = argparse.ArgumentParser(
        description="Plan a raw prompt first, then optionally execute the plan with Codex."
    )
    parser.add_argument("prompt", nargs="*", help="Raw prompt. Reads stdin if omitted.")
    parser.add_argument("--execute", action="store_true", help="Run codex exec with the plan.")
    parser.add_argument("--cwd", default=os.getcwd(), help="Workspace for codex exec.")
    parser.add_argument("--max-tasks", type=int, default=8)
    parser.add_argument(
        "--max-rounds",
        type=int,
        default=3,
        help="Maximum Codex execute -> Gate follow-up rounds before returning incomplete.",
    )
    parser.add_argument(
        "--codex-arg",
        action="append",
        default=[],
        help="Extra argument passed to codex exec. Repeat for multiple args.",
    )
    args = parser.parse_args(argv)

    prompt = " ".join(args.prompt).strip() or sys.stdin.read().strip()
    gate = CodexGate(
        planner=TaskPlanner(max_tasks=args.max_tasks),
        followup_planner=FollowupPlanner(max_tasks=args.max_tasks),
    )
    result = gate.run(
        prompt,
        execute=args.execute,
        cwd=args.cwd,
        codex_args=args.codex_arg,
        max_rounds=args.max_rounds,
    )
    stream = sys.stdout if result.exit_code == 0 else sys.stderr
    if result.output:
        print(result.output, file=stream)
    return result.exit_code


if __name__ == "__main__":
    raise SystemExit(main())

File: plugins/task-gate/scripts/codex_gate.ts
#!/usr/bin/env -S node --experimental-strip-types
import { spawnSync } from "node:child_process";
import { readFileSync } from "node:fs";
import { fileURLToPath } from "node:url";

import {
  PLAN_SCHEMA,
  PlanError,
  Task,
  TaskPlan,
  TaskPlanner,
  buildDefaultThinker,
  stringList,
  taskFromRaw,
  tryParseJson,
  type Thinker,
} from "./task_gate.ts";

type JsonValue = unknown;

type RunResult = {
  stdout?: string;
  stderr?: string;
  status?: number | null;
  error?: Error & { code?: string };
};

type Runner = (args: string[], options: Record<string, JsonValue>) => RunResult;

export const FOLLOWUP_SCHEMA = {
  type: "object",
  properties: {
    complete: { type: "boolean" },
    summary: { type: "string" },
    next_tasks: PLAN_SCHEMA.properties.tasks,
  },
  required: ["complete", "summary"],
  additionalProperties: false,
};

export class GateResult {
  exitCode: number;
  output: string;

  constructor({ exitCode, output = "" }: { exitCode: number; output?: string }) {
    this.exitCode = exitCode;
    this.output = output;
  }
}

export class CodexRunResult {
  exitCode: number;
  output: string;

  constructor({ exitCode, output = "" }: { exitCode: number; output?: string }) {
    this.exitCode = exitCode;
    this.output = output;
  }
}

export class FollowupDecision {
  complete: boolean;
  summary: string;
  nextTasks: Task[];

  constructor({
    complete,
    summary = "",
    nextTasks = [],
  }: {
    complete: boolean;
    summary?: string;
    nextTasks?: Task[];
  }) {
    this.complete = complete;
    this.summary = summary;
    this.nextTasks = nextTasks;
  }
}

function defaultRunner(args: string[]): RunResult {
  const [command, ...rest] = args;
  if (!command) return { status: 1, stderr: "missing command" };
  return spawnSync(command, rest, { encoding: "utf8" }) as RunResult;
}

export class CodexExecutor {
  command: string;
  runner: Runner;

  constructor({ command, runner = defaultRunner }: { command?: string; runner?: Runner } = {}) {
    this.command = command || process.env.TASK_GATE_CODEX_BIN || "codex";
    this.runner = runner;
  }

  execute({
    codexPrompt,
    cwd,
    extraArgs = [],
  }: {
    codexPrompt: string;
    cwd?: string;
    extraArgs?: string[];
  }): CodexRunResult {
    const args = [this.command, "exec", "--skip-git-repo-check"];
    if (cwd) args.push("-C", cwd);
    args.push(...extraArgs, codexPrompt);
    const completed = this.runner(args, { text: true, check: false, captureOutput: true });
    if (completed.error?.code === "ENOENT") {
      throw new PlanError(`codex command not found: ${this.command}`);
    }
    const stdout = (completed.stdout || "").trim();
    const stderr = (completed.stderr || "").trim();
    return new CodexRunResult({
      exitCode: Number(completed.status ?? 0),
      output: [stdout, stderr].filter(Boolean).join("\n"),
    });
  }
}

export class FollowupPlanner {
  thinker: Thinker;
  maxTasks: number;

  constructor({ thinker, maxTasks = 8 }: { thinker?: Thinker; maxTasks?: number } = {}) {
    this.thinker = thinker || buildDefaultThinker(FOLLOWUP_SCHEMA);
    this.maxTasks = maxTasks;
  }

  assess({
    plan,
    codexOutput,
    exitCode,
    roundNumber,
  }: {
    plan: TaskPlan;
    codexOutput: string;
    exitCode: number;
    roundNumber: number;
  }): FollowupDecision {
    const prompt = buildFollowupPrompt({
      plan,
      codexOutput,
      exitCode,
      roundNumber,
      maxTasks: this.maxTasks,
    });
    return parseFollowupOutput(this.thinker.think(prompt), this.maxTasks);
  }
}

export class CodexGate {
  planner: { plan(prompt: string): TaskPlan };
  executor: CodexExecutor;
  followupPlanner: { assess(args: Record<string, JsonValue>): FollowupDecision };

  constructor({
    planner,
    executor,
    followupPlanner,
  }: {
    planner?: { plan(prompt: string): TaskPlan };
    executor?: CodexExecutor;
    followupPlanner?: { assess(args: Record<string, JsonValue>): FollowupDecision };
  } = {}) {
    this.planner = planner || new TaskPlanner();
    this.executor = executor || new CodexExecutor();
    this.followupPlanner = followupPlanner || new FollowupPlanner();
  }

  run({
    prompt,
    execute = false,
    cwd,
    codexArgs = [],
    maxRounds = 3,
  }: {
    prompt: string;
    execute?: boolean;
    cwd?: string;
    codexArgs?: string[];
    maxRounds?: number;
  }): GateResult {
    let plan: TaskPlan;
    try {
      plan = this.planner.plan(prompt);
    } catch (error) {
      if (error instanceof PlanError) {
        return new GateResult({ exitCode: 1, output: `task-gate: ${error.message}` });
      }
      throw error;
    }

    if (!execute) return new GateResult({ exitCode: 0, output: plan.asNumberedText() });
    if (maxRounds < 1) {
      return new GateResult({ exitCode: 1, output: "task-gate: max_rounds must be at least 1" });
    }

    let currentPlan = plan;
    const reports = ["Initial task plan:", plan.asNumberedText()];
    for (let roundNumber = 1; roundNumber <= maxRounds; roundNumber += 1) {
      const codexPrompt = buildCodexExecutionPrompt(currentPlan, roundNumber);
      try {
        const runResult = this.executor.execute({
          codexPrompt,
          cwd,
          extraArgs: codexArgs,
        });
        const decision = this.followupPlanner.assess({
          plan: currentPlan,
          codexOutput: runResult.output,
          exitCode: runResult.exitCode,
          roundNumber,
        });
        reports.push(formatRoundReport(roundNumber, runResult, decision));
        if (decision.complete && runResult.exitCode === 0) {
          reports.push("Task Gate completion verdict: complete.");
          return new GateResult({ exitCode: 0, output: reports.join("\n\n") });
        }
        if (decision.complete && runResult.exitCode !== 0) {
          reports.push(
            "Task Gate marked the work complete, but Codex exited nonzero; continuing because completion is not proven.",
          );
        }
        if (!decision.nextTasks.length) {
          reports.push("task-gate: task is not complete and Gate returned no next tasks.");
          return new GateResult({ exitCode: 1, output: reports.join("\n\n") });
        }
        currentPlan = new TaskPlan({
          sourcePrompt: `Gate follow-up after execution round ${roundNumber}`,
          tasks: decision.nextTasks,
        });
      } catch (error) {
        if (error instanceof PlanError) {
          reports.push(`task-gate: ${error.message}`);
          return new GateResult({ exitCode: 1, output: reports.join("\n\n") });
        }
        throw error;
      }
    }

    reports.push(
      "task-gate: max execution rounds reached before completion; refusing to report success while work remains.",
    );
    return new GateResult({ exitCode: 1, output: reports.join("\n\n") });
  }
}

export function buildCodexExecutionPrompt(plan: TaskPlan, roundNumber = 1): string {
  return (
    "Task Gate has already converted the user's raw request into the authorized task plan below. " +
    "Execute only these numbered tasks in order.\n" +
    `This is execution round ${roundNumber}.\n\n` +
    `${formatExecutionPlan(plan)}\n\n` +
    "Execution rules:\n" +
    "- Treat the numbered plan as the task boundary.\n" +
    "- Do not reinterpret the original raw request; it is intentionally absent.\n" +
    "- If a required task is missing or unsafe, stop and report the blocker.\n" +
    "- Before claiming completion, run suitable verification and report evidence.\n" +
    "- At the end of your response, always include a Detailed completion summary.\n\n" +
    "Detailed completion summary requirements:\n" +
    "- Work completed: list the concrete tasks handled and files or commands involved.\n" +
    "- Verification: list commands, checks, and exact results, or say why they were not run.\n" +
    "- Remaining work: list any unfinished task, missing verification, or uncertainty.\n" +
    "- Blockers: list blockers or write none.\n" +
    "- Completion verdict: write complete only when all authorized tasks and verification are done; otherwise write incomplete.\n" +
    "This detailed summary is mandatory because Task Gate will read it to decide the next tasks."
  );
}

export function buildFollowupPrompt({
  plan,
  codexOutput,
  exitCode,
  roundNumber,
  maxTasks = 8,
}: {
  plan: TaskPlan;
  codexOutput: string;
  exitCode: number;
  roundNumber: number;
  maxTasks?: number;
}): string {
  return (
    "You are Task Gate reviewing the end of a Codex execution round. " +
    "The raw user prompt is intentionally absent; use only the authorized task plan and Codex's detailed completion summary.\n" +
    "Return only JSON matching this shape: " +
    '{"complete":false,"summary":"detailed assessment",' +
    '"next_tasks":[{"title":"next task","details":"optional detail",' +
    '"acceptance_criteria":["optional check"]}]}.\n' +
    `If complete is false, provide 1 to ${maxTasks} concrete next_tasks so Codex can continue. ` +
    "If complete is true, next_tasks may be empty. Do not mark complete when verification is missing, " +
    "Codex exited nonzero, or the summary says work remains.\n\n" +
    `Execution round: ${roundNumber}\n` +
    `Codex exit code: ${exitCode}\n\n` +
    "Authorized task plan:\n" +
    `${formatExecutionPlan(plan)}\n\n` +
    "Codex detailed completion summary and output:\n" +
    clipText(codexOutput)
  );
}

export function parseFollowupOutput(output: string, maxTasks = 8): FollowupDecision {
  const parsed = tryParseJson(output);
  if (parsed === null) throw new PlanError("follow-up gate returned invalid JSON");
  return followupFromJson(parsed, maxTasks);
}

function followupFromJson(parsed: JsonValue, maxTasks: number): FollowupDecision {
  const record = asRecord(parsed);
  if (record && asRecord(record.structured_output)) return followupFromJson(record.structured_output, maxTasks);
  if (record && typeof record.result === "string") {
    const nested = tryParseJson(record.result);
    if (nested !== null) return followupFromJson(nested, maxTasks);
  }
  if (!record) throw new PlanError("follow-up gate JSON must be an object");
  const complete = boolFromRaw(
    Object.hasOwn(record, "complete") ? record.complete : record.is_complete || record.done,
  );
  const summary = String(record.summary || record.assessment || "").trim();
  const rawTasks = record.next_tasks || record.tasks || [];
  const nextTasks: Task[] = [];
  if (Array.isArray(rawTasks)) {
    for (const raw of rawTasks.slice(0, maxTasks)) {
      const task = taskFromRaw(raw, nextTasks.length + 1);
      if (task) nextTasks.push(task);
    }
  }
  return new FollowupDecision({ complete, summary, nextTasks });
}

function boolFromRaw(raw: JsonValue): boolean {
  if (typeof raw === "boolean") return raw;
  if (typeof raw === "string") {
    const normalized = raw.trim().toLowerCase();
    if (["true", "yes", "done", "complete", "completed"].includes(normalized)) return true;
    if (["false", "no", "incomplete", "remaining", "blocked"].includes(normalized)) return false;
  }
  throw new PlanError('follow-up gate JSON must contain boolean "complete"');
}

function clipText(text: string): string {
  const limit = Number(process.env.TASK_GATE_CODEX_OUTPUT_CHARS || "12000");
  if (text.length <= limit) return text;
  const omitted = text.length - limit;
  return `[omitted ${omitted} earlier chars]\n${text.slice(-limit)}`;
}

function formatRoundReport(roundNumber: number, runResult: CodexRunResult, decision: FollowupDecision): string {
  const lines = [
    `Execution round ${roundNumber}`,
    `Codex exit code: ${runResult.exitCode}`,
    "Codex detailed completion summary:",
    indent(runResult.output || "<no output>"),
    "Gate follow-up:",
    indent(decision.summary || "<no summary>"),
  ];
  if (decision.nextTasks.length) {
    lines.push("Gate next tasks:");
    lines.push(indent(new TaskPlan({ sourcePrompt: "", tasks: decision.nextTasks }).asNumberedText()));
  }
  return lines.join("\n");
}

function indent(text: string): string {
  return text
    .split(/\r?\n/)
    .map((line) => (line ? `  ${line}` : ""))
    .join("\n");
}

function formatExecutionPlan(plan: TaskPlan): string {
  const lines: string[] = [];
  for (const task of plan.tasks) {
    lines.push(`${task.id}. ${task.title}`);
    if (task.details) lines.push(`   Details: ${task.details}`);
    if (task.acceptanceCriteria.length) {
      lines.push("   Acceptance criteria:");
      lines.push(...task.acceptanceCriteria.map((criterion) => `   - ${criterion}`));
    }
  }
  return lines.join("\n");
}

function asRecord(value: JsonValue): Record<string, JsonValue> | null {
  return value && typeof value === "object" && !Array.isArray(value) ? (value as Record<string, JsonValue>) : null;
}

type CliArgs = {
  prompt: string[];
  execute: boolean;
  cwd: string;
  maxTasks: number;
  maxRounds: number;
  codexArgs: string[];
};

function parseArgs(argv: string[]): CliArgs {
  const args: CliArgs = {
    prompt: [],
    execute: false,
    cwd: process.cwd(),
    maxTasks: 8,
    maxRounds: 3,
    codexArgs: [],
  };
  for (let i = 0; i < argv.length; i += 1) {
    const item = argv[i];
    if (item === "--execute") args.execute = true;
    else if (item === "--cwd") args.cwd = argv[++i] || process.cwd();
    else if (item === "--max-tasks") args.maxTasks = Number(argv[++i] || "8");
    else if (item === "--max-rounds") args.maxRounds = Number(argv[++i] || "3");
    else if (item === "--codex-arg") args.codexArgs.push(argv[++i] || "");
    else if (item === "-h" || item === "--help") {
      printHelp();
      process.exit(0);
    } else if (item) args.prompt.push(item);
  }
  return args;
}

function printHelp(): void {
  console.log(`usage: codex_gate.ts [--execute] [--cwd DIR] [--max-tasks N] [--max-rounds N] [--codex-arg ARG] [prompt ...]

Plan a raw prompt first, then optionally execute the plan with Codex.`);
}

export function main(argv = process.argv.slice(2)): number {
  const args = parseArgs(argv);
  const prompt = args.prompt.join(" ").trim() || readFileSync(0, "utf8").trim();
  const gate = new CodexGate({
    planner: new TaskPlanner({ maxTasks: args.maxTasks }),
    followupPlanner: new FollowupPlanner({ maxTasks: args.maxTasks }),
  });
  const result = gate.run({
    prompt,
    execute: args.execute,
    cwd: args.cwd,
    codexArgs: args.codexArgs.filter(Boolean),
    maxRounds: args.maxRounds,
  });
  if (result.output) {
    const stream = result.exitCode === 0 ? process.stdout : process.stderr;
    stream.write(`${result.output}\n`);
  }
  return result.exitCode;
}

if (process.argv[1] === fileURLToPath(import.meta.url)) {
  process.exitCode = main();
}
